LSPS5/README.md

```diff
@@ -0,0 +1,375 @@
+before causing the message to be displayed to the specified user.
+If the user then taps on this "push notification" on their screen,
+the application is awoken and it can then re-establish its TCP
+connections.
+
+There may be multiple implementations of mobile LSPS-supporting
+clients, each of which run their own application developer server
+and have their own signing keys for these push notifications.
+
+This specification provides a way for a mobile client to
+register some specific webhook by which the LSP can signal a
+push notification to the application developer server, which
+will in turn convert the push notification to one that it itself
+signs and can send to the mobile OS developer server.
+
+### Actors
+
+The 'client' is a specific instance of the client application
+running on some mobile device.
+
+The 'LSP' is a published node that may need to awaken some
+client instance in order to handle some event.
+
+The 'notification delivery service' is the server that a mobile
+client application developer maintains, which is contacted by
+the LSP via an HTTPS `POST` request, in order for the LSP
+to awaken some client instance.
+
+The 'mobile OS server' is the entity that actually allows
+push notifications to be sent to actual mobile devices.
+
+## Protocol
+
+### Webhook Registration API
+
+The client can register a specific webhook for the LSP to
+contact in case the client is asleep, but the LSP needs the
+client to be online and ready to interact.
+
+The client uses the `lsps5.set_webhook` call to specify the
+HTTPS URI that the LSP should `POST` in order to send a
+push notification to the client user.
+It takes the parameters below:
+
+```JSON
+{
+    "webhook": "https://www.example.org/push?l=1234567890abcdefghijklmnopqrstuv&c=best"
+}
+```
+
+`webhook` is a required string, containing the URI of the
+webhook that the LSP can use to push a notification to the
+client.
+
+The client:
+
+* MUST use a HTTPS-protocol URI to the notification delivery
+  service.
+* MUST encode the following data in any part of the URI
+  (hostname, path, or `GET` parameters):
+  * The LSP node ID.
+  * Some authorization to allow the given LSP to wake the
+    client, ideally a proof that the client has a channel
+    or other relationship with the LSP.
+  * Some way to identify the mobile device the client is
+    running on.
+* MAY encode any additional information in the URI.
+
+> **Rationale** By placing most data in the URI, the LSP
+> interface is simplified and the client developer has the
+> flexibility to add any additional information they want
+> the LSP to simply pass through from the client instance
+> to the notification delivery service.
+
+The LSP:
+
+* MUST persist the given webhook and associate it with the
+  client node ID before responding to this call.
+* MUST use the `webhook` specified in the latest
+  `lsps5.set_webhook` call.
+* MAY forget the `webhook` after 7 days (7 x 24 hours) from
```

Comment:
The previous requirement says "MUST use the `webhook` specified in the latest `lsps5.set_webhook` call.", which implies that the LSP has no need to store historical `webhook`s.  Do you really want me to explicitly say "if there is an existing registered `webhook`, the LSP MUST replace the existing `webhook` with the new call" somewhere?

Reply:
I missed that part. Reading again I see the spec allows actually one webhook per node id which won't work for the case where user use the node from multiple devices. I really think the LSP should allow registering multiple urls per node and change the `set_webhook` to `add_webhook` Otherwise different apps will override each other.

Reply:
If there is an `add_webhook` should there be a `remove_webhook` as well?

Would a `set_webhook` that accepts an array of `webhooks` work?

What are the expected semantics of multiple webhooks? LSP calls all of them? If so, can this be used by a client to cause the LSP to inadvertently mount a DDoS by adding tons of links to a victim website and then triggering a single payment that it fails?

If multiple webhooks are supported, I think we should have some way of limiting how many webhooks maximum can be registered by a single client.

Reply:
Bikeshed proposal:

`add_webhook {webhook: "blah"}` => `{"num_existing_webhooks`: 42, "num_max_webhooks"": 44, "webhook_id": 33}`

`remove_webhook {"webhook_id": 33}` => `{"num_existing_webhooks": 41, "num_max_webhooks": 44}`

`list_webhooks {}` => `{"webhooks": [ {"webhook_id": 33, "webhook": blah}, {"webhook_id": 32, "webhook": "blah blah"} ]}`

?

Reply:
I like this proposal. I think it simple and flexible.
Also think lsp MAY remove webhooks if not updated for a while (1 week?).

Reply:
Unresolved yet, would like further discussion on this both here and on Telegram.

Reply:
multi-webhook done.